_posts/2022-10-07-arrow-parquet-encoding-part-2.md

```diff
@@ -0,0 +1,344 @@
+
+For example, a list with offsets `[0, 2, 3, 3]` contains 3 pairs of offsets, `(0,2)`, `(2,3)`, and `(3,3)`, and therefore represents a `ListArray` of length 3 with the following values:
+
+```text
+0: [child[0], child[1]]
+1: []
+2: [child[2]]
+```
+
+For the example above with 4 JSON documents, this would be encoded in Arrow as
+
+
+```text
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+                          ┌──────────────────┐ │
+│    ┌─────┐   ┌─────┐    │ ┌─────┐   ┌─────┐│
+     │  1  │   │  0  │    │ │  1  │   │  1  ││ │
+│    ├─────┤   ├─────┤    │ ├─────┤   ├─────┤│
+     │  0  │   │  1  │    │ │  0  │   │ ??  ││ │
+│    ├─────┤   ├─────┤    │ ├─────┤   ├─────┤│
+     │  1  │   │  1  │    │ │  1  │   │  2  ││ │
+│    ├─────┤   ├─────┤    │ └─────┘   └─────┘│
+     │  1  │   │  1  │    │ Validity   Values│ │
+│    └─────┘   ├─────┤    │                  │
+               │  3  │    │ child[0]         │ │
+│    Validity  └─────┘    │ PrimitiveArray   │
+                          │                  │ │
+│              Offsets    └──────────────────┘
+     "a"                                       │
+│    ListArray
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
+
+More technical detail is available in the [ListArray format specification](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout).
+
+
+### Parquet Repetition Levels
+
+The example above with 4 JSON documents can be stored in this Parquet schema
+
+```text
+message schema {
+  optional group a (LIST) {
+    repeated group list {
+      optional int32 element;
+    }
+  }
+}
+```
+
+In order to encode lists, Parquet stores an integer *repetition level* in addition to a definition level. A repetition level identifies where in the hierarchy of repeated fields the current value is to be inserted. A value of `0` means a new list in the top-most repeated list, a value of `1` means a new element within the top-most repeated list, a value of `2` means a new element within the second top-most repeated list, and so on.
+
+*Protip*: for the topmost level list, the number of zeros in the `repetition` levels must match the number of rows.
+
+Each repeated field also has a corresponding definition level, however, in this case rather than indicating a null value, they indicate an empty array.
+
+
+```text
+┌─────────────────────────────────────┐
+│  ┌─────┐      ┌─────┐               │
+│  │  3  │      │  0  │               │
+│  ├─────┤      ├─────┤               │
+│  │  0  │      │  0  │               │
+│  ├─────┤      ├─────┤      ┌─────┐  │
+│  │  1  │      │  0  │      │  1  │  │
+│  ├─────┤      ├─────┤      ├─────┤  │
+│  │  2  │      │  0  │      │  2  │  │
+│  ├─────┤      ├─────┤      └─────┘  │
+│  │  3  │      │  1  │               │
+│  └─────┘      └─────┘               │
```

Comment:
@tustvold I don't think this example is correct (I think I mis translated it / mashed it up incorrectly). Among other issues, it has 5 values in definition/repetition levels but there are only 4 documents 🤔

Reply:
But there are 5 levels, as one of the arrays has two elements?

Reply:
Awesome -- I will study it more carefully; Good thing there is a blog about this stuff I can read 😆

Reply:
Double checked with

```
let mut list_builder = ListBuilder::new(Int32Builder::new());
        list_builder.values().append_value(1);
        list_builder.append(true);
        list_builder.append(false);
        list_builder.append(true);
        list_builder.values().append_null();
        list_builder.values().append_value(2);
        list_builder.append(true);
        let values = Arc::new(list_builder.finish()) as ArrayRef;

        let mut builder = LevelInfoBuilder::try_new(
            &Field::new("test", values.data_type().clone(), true),
            Default::default(),
        )
        .unwrap();
        builder.write(&values, 0..4);
        let levels = builder.finish();

        assert_eq!(levels.len(), 1);

        let list_level = levels.get(0).unwrap();

        let expected_level = LevelInfo {
            def_levels: Some(vec![3, 0, 1, 2, 3]),
            rep_levels: Some(vec![0, 0, 0, 0, 1]),
            non_null_indices: vec![0, 2],
            max_def_level: 3,
            max_rep_level: 1,
        };
        assert_eq!(list_level, &expected_level);
```

Reply:
Thank you -- I was losing sleep over this particular example